File: common/brownshome/scriptwars/server/game/tanks/Direction.java
package brownshome.scriptwars.server.game.tanks;

import brownshome.scriptwars.client.tankapi.Coordinates;

public enum Direction {
	UP(0, -1),
	LEFT(-1, 0),
	DOWN(0, 1),
	RIGHT(1, 0);
	
	final int dx, dy;
	
	Direction(int dx, int dy) {
		this.dx = dx;
		this.dy = dy;
	}
	
	public int moveX(int x) {
		return x + dx;
	}
	
	public int moveY(int y) {
		return y + dy;
	}
	
	public Coordinates move(Coordinates coord) {
		return new Coordinates(moveX(coord.getX()), moveY(coord.getY()));
	}
	
	public Direction opposite() {
		return values()[(ordinal() + 2) % 4];
	}
	
	public Direction clockwise() {
		return values()[(ordinal() + 3) % 4];
	}
	
	public Direction antiClockwise() {
		return values()[(ordinal() + 1) % 4];
	}
	
	/**
	 * @return A direction object represented by the given coordinates. If the coordinates
	 * do not match a cardinal direction {@code null} is returned.
	 */
	public static Direction getDirection(int x, int y) {
		if(x > 0 && y == 0) return RIGHT;
		if(x < 0 && y == 0) return LEFT;
		if(x == 0 && y < 0) return UP;
		if(x == 0 && y > 0) return DOWN;
		return null;
	}
}

File: client/python/Direction.py
from Coordinates import *

class Direction:
    def __init__(self, dx, dy, code):
        self.dx = dx
        self.dy = dy
        self.code = code

    def move(self, coord):
        c = Coordinates()
        c.x = coord.x + self.dx
        c.y = coord.y + self.dy
        return c

    def opposite(self):
        return Direction.values[(self.code + 2) % 4]

    def clockwise(self):
        return Direction.values[(self.code + 3) % 4]

    def anitClockwise(self):
        return Direction.values[(self.code + 1) % 4]

    def getDirection(dx, dy):
        if(x > 0 and y == 0):
            return Direction.RIGHT
        if(x < 0 and y == 0):
            return Direction.LEFT
        if(x == 0 and y < 0):
            return Direction.UP
        if(x == 0 and y > 0):
            return Direction.DOWN

        return None;

    def getDirection(toPos, fromPos):
        return getDirection(toPos.x - fromPos.x, toPos.y - fromPos.y)

Direction.UP = Direction(0, -1, 0)
Direction.LEFT = Direction(-1, 0, 1)        
Direction.DOWN = Direction(0, 1, 2)
Direction.RIGHT = Direction(1, 0, 3)

Direction.values = [Direction.UP, Direction.LEFT, Direction.DOWN, Direction.RIGHT]
